### ISSUE TITLE
Serialisation of Set crashes

### ISSUE THREAD

**Issue by Ghostbird** [sebastienros/jint#1075]
Running this code crashes:
```csharp
JsonConvert.SerializeObject(new Engine().Evaluate("new Set()"));
```
This is the start of the stacktrace:
```
Exception thrown: 'Newtonsoft.Json.JsonSerializationException' in Newtonsoft.Json.dll: 'Self referencing loop detected for property 'Engine' with type 'Jint.Engine'. Path 'Engine.Realm.Intrinsics.Object.PrototypeObject'.'
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.CheckForCircularReference(JsonWriter writer, Object value, JsonProperty property, JsonContract contract, JsonContainerContract containerContract, JsonProperty containerProperty)
```

Is this an error in Jint? None of the other prototype containing objects (e.g. object, Array) seem to throw this exception.

**Comment by lahma** [comment]
Probably a bug. Would you like to offer a PR with a test case? 🙂

**Comment by Ghostbird** [comment]
See #1078

**Comment by lahma** [comment]
After a brief look into this I believe the problem is that `ObjectInstance` and its inheritors don' provide any iterable interface for Json.NET: https://www.newtonsoft.com/json/help/html/SerializationGuide.htm

Not sure what's the best way to tackle this as `ObjectInstance` is a key-value kind of thing and some of it's logical inheritors are value iterable, like `Set` and `Array`. Currently serialization works for `ArrayInstance` as it has explicit implementation for `IEnumerable` unlike other things in Jint.

**Comment by lahma** [comment]
Easy fix would be to change `ObjectInstance`'s `Engine` property to `protected internal` (which it should probably be anyway), but it won't make the JSON output follow the JS semantics. 

@Ghostbird where does the need to get Jint object instances serialized following the JS logic (`JSON.stringify`) originate from? Just trying to understand the underlying problem.

**Comment by Ghostbird** [comment]
I'm not sure whether this should be treated as a bug, since it turns out a Set isn't natively serialisable in Javascript anyway. I never realised that when I wrote this issue.
```javascript
{
  const x = new Set();
  x.add('a');
  x.add('b');
  x.add('c');
  JSON.stringify(x);
}
```
The result of the last expression is:
```json
"{}"
```
However if I replace the last expression with 
```javascript
  JSON.stringify(Array.from(x));
```
The result becomes:
```json
"[\"a\",\"b\",\"c\"]"
```

Hence I conclude:
- In Javascript, direct serialisation of a set produces no useful information.
- In C# serialisation of the data that represents Jint's implementation of a Javascript set provides no useful information
- In both cases the information I desired to get is accessible by wrapping the set in an `Array.from` call.

I now kind of prefer the C# situation where it explicitly fails, instead of just silently returning an empty object. Even if the C# error is thrown "for the wrong reasons". If C# had actually correctly serialised an empty object, we'd have had a much sneakier bug in our software.

I no longer think this is a bug in Jint.
